### PR TITLE
Add enabled/disabled in states

### PR DESCRIPTION
**Pull request by LucasZdv**
Hello, I am pleased to make this small contribution.

- It allows you to enable / disable states individually and in bulk.
- Enable the search box
- Allows filtering by enabled

screenshot-dev-integrity gotech link-2020 11 20-12_18_36

### CODE REVIEW

File: controllers/Locations.php [hunk 1]
```diff
@@ -80,7 +92,7 @@ public function onDisableLocations()
             Flash::success(Lang::get('rainlab.location::lang.locations.disable_success'));
         }
 
-        return Backend::redirect('rainlab/location/locations');
+        return Redirect::to(Request::url());
```
Comment (LukeTowers) [inline controllers/Locations.php:95]:
```suggestion
        return redirect()->refresh();
```

File: models/State.php [hunk 2]
```diff
@@ -66,7 +66,12 @@ public static function getNameList($countryId)
             return self::$nameList[$countryId];
         }
 
-        return self::$nameList[$countryId] = self::whereCountryId($countryId)->orderBy('name', 'asc')->lists('name', 'id');
+        return self::$nameList[$countryId] = self::whereCountryId($countryId)->IsEnabled()->orderBy('name', 'asc')->lists('name', 'id');
```
Comment (LukeTowers) [inline models/State.php:69]:
```suggestion
        return self::$nameList[$countryId] = self::whereCountryId($countryId)->isEnabled()->orderBy('name', 'asc')->lists('name', 'id');
```

File: models/state/filters.yaml [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+        # Filter name
+        label: rainlab.location::lang.locations.disabled_label
+
+        # Filter type
+        type: switch
+
+        # SQL Conditions
+        conditions:
+            - is_enabled = '1'
+            - is_enabled = '0'
```
Comment (LukeTowers) [inline models/state/filters.yaml:13]:
```suggestion
    enabled:
        label: rainlab.location::lang.locations.disabled_label
        type: switch
        scope: isEnabled
```

File: updates/version.yaml [hunk 4]
```diff
@@ -23,3 +23,6 @@
 1.1.5:
     - Include Italian states (province)
     - seed_it_states.php
+1.1.6:
+    - Add enabled/disabled in states. Don't worry, default is true.
```
Comment (LukeTowers) [inline updates/version.yaml:27]:
```suggestion
    - Added ability to disable specific states
```

File: updates/add_enabled_states.php [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+use RainLab\Location\Models\Country;
+use October\Rain\Database\Schema\Blueprint;
+use October\Rain\Database\Updates\Migration;
+
+class AddEnabledStates extends Migration
+{
+    public function up()
+    {
+        Schema::table('rainlab_location_states', function(Blueprint $table) {
+            $table->boolean('is_enabled')->default(true);
```
Comment (LukeTowers) [inline updates/add_enabled_states.php:13]:
```suggestion
            $table->boolean('is_enabled')->after('country_id')->default(true);
```

File: controllers/locations/config_relation.yaml [hunk 6]
```diff
@@ -7,4 +7,6 @@ states:
     list: ~/plugins/rainlab/location/models/state/columns.yaml
     form: ~/plugins/rainlab/location/models/state/fields.yaml
     view:
-        toolbarButtons: create|delete
+        showSearch: true
+        toolbarPartial: ~/plugins/rainlab/location/controllers/locations/_states_list_toolbar.htm
+        filter: ~/plugins/rainlab/location/models/state/filters.yaml
```
Comment (LukeTowers) [inline controllers/locations/config_relation.yaml:12]:
```suggestion
        toolbarPartial: $/rainlab/location/controllers/locations/_states_list_toolbar.htm
        filter: $/rainlab/location/models/state/filters.yaml
```

File: controllers/locations/_states_list_toolbar.htm [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        class="btn btn-sm btn-secondary oc-icon-trash-o control-disabled" 
+        data-request="onRelationButtonDelete" 
+        data-request-confirm="Are you sure?" 
+        data-request-success="$.oc.relationBehavior.changed('states', 'deleted')" 
+        data-trigger-action="enable" 
+        data-trigger="#Locations-update-RelationController-states-view .control-list input[type=checkbox]" 
+        data-trigger-condition="checked" 
+        data-stripe-load-indicator=""
+        disabled="">
+        Delete
```
Comment (LukeTowers) [inline controllers/locations/_states_list_toolbar.htm:39]: Need translation here

File: controllers/locations/_states_list_toolbar.htm [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+        data-trigger-condition="checked"
+        data-stripe-load-indicator>
+        <?= e(trans('rainlab.location::lang.locations.enable_or_disable')) ?>
+    </button>
+
+    <button 
+        onclick="$(this).data('request-data', {checked: $('#Locations-update-RelationController-states-view .control-list').listWidget('getChecked') })" 
+        class="btn btn-sm btn-secondary oc-icon-trash-o control-disabled" 
+        data-request="onRelationButtonDelete" 
+        data-request-confirm="Are you sure?"
```
Comment (LukeTowers) [inline controllers/locations/_states_list_toolbar.htm:32]: Need translation here

File: controllers/locations/_states_list_toolbar.htm [hunk 9]
```diff
@@ -0,0 +1,41 @@
+<div data-control="toolbar" data-disposable="">
+
+    <a 
+        data-handler="onRelationButtonCreate" 
+        data-control="popup" 
+        data-size="huge" 
+        href="javascript:;" 
+        class="btn btn-sm btn-secondary oc-icon-file">
+        Create State
```
Comment (LukeTowers) [inline controllers/locations/_states_list_toolbar.htm:9]: Need translation here

File: controllers/Locations.php [hunk 10]
```diff
@@ -63,8 +66,17 @@ public function onDisableLocations()
...
+                if(post('location_type') == 'country') {
+                    if (!$object = Country::find($objectId)) {
+                        continue;
+                    }
+                }
+                
+                if(post('location_type') == 'state') {
+                    if (!$object = State::find($objectId)) {
+                        continue;
+                    }
```
Comment (LukeTowers) [inline controllers/Locations.php:79]:
```suggestion
                $object = null;
                switch (post('location_type')) {
                    case 'country':
                        $object = Country::find($objectId);
                        break;
                    case 'state':
                        $object = State::find($objectId);
                        break;
                }
                
                if (!$object) {
                    continue;
                }
```

**Review by LukeTowers — CHANGES_REQUESTED** [review]

File: controllers/Locations.php [hunk 11]
```diff
@@ -4,8 +4,10 @@
 use Flash;
 use Backend;
 use Redirect;
+use Request;
```
Comment (LukeTowers) [inline controllers/Locations.php:7]:
```suggestion
use Flash;
use Backend;
```

File: controllers/Locations.php [hunk 12]
```diff
@@ -63,9 +64,20 @@ public function onDisableLocations()
...
+                        break;
+                    case 'state':
+                        $object = State::find($objectId);
+                        break;
+                }
+                
+                if (!$object) {
                     continue;
                 }
+                }
```
Comment (LukeTowers) [inline controllers/Locations.php:80]:
```suggestion
```

File: controllers/locations/_states_list_toolbar.htm (hunk 7 above)
Comment (LucasZdv) [reply controllers/locations/_states_list_toolbar.htm]: https://github.com/rainlab/location-plugin/commit/19212d1b148380e4197f9eaafabbad9a7fbc556d

File: controllers/locations/_states_list_toolbar.htm (hunk 8 above)
Comment (LucasZdv) [reply controllers/locations/_states_list_toolbar.htm]: https://github.com/rainlab/location-plugin/commit/19212d1b148380e4197f9eaafabbad9a7fbc556d

File: controllers/locations/_states_list_toolbar.htm (hunk 9 above)
Comment (LucasZdv) [reply controllers/locations/_states_list_toolbar.htm]: https://github.com/rainlab/location-plugin/commit/19212d1b148380e4197f9eaafabbad9a7fbc556d

**Comment by LucasZdv** [comment]
@LukeTowers Thanks for everything. 

I want to add this in the states, any idea how i can it? - 

```
   public function listInjectRowClass($record, $definition = null)
    {
        if (!$record->is_enabled) {
            return 'safe disabled';
        }
    }
```

This adds a css class to the disabled countries, I want to do the same in the states to make it verbose.

File: controllers/locations/_states_list_toolbar.htm [hunk 13]
```diff
@@ -6,7 +6,7 @@
         data-size="huge" 
         href="javascript:;" 
         class="btn btn-sm btn-secondary oc-icon-file">
-        Create State
+        <?= e(trans('backend::lang.form.create')) ?> <?= e(trans('rainlab.location::lang.state.label')) ?>
```
Comment (LukeTowers) [inline controllers/locations/_states_list_toolbar.htm:9]:
```suggestion
        <?= e(trans('backend::lang.form.create_name', ['name' => trans('rainlab.location::lang.state.label')])) ?>
```

**Comment by LukeTowers** [comment]
@LucasZdv you'd have to extend the relation view widget and bind into the list.injectRowClass event. Something like

```php
public function relatedExtendViewWidget($widget)
{
    $widget->bindEvent('list.injectRowClass', function ($record) {
        return !$record->is_enabled ? 'safe disabled';
    }
}
```

File: controllers/locations/_states_list_toolbar.htm (hunk 13 above)
Comment (LucasZdv) [reply controllers/locations/_states_list_toolbar.htm]: backend::lang.form.create_name does not exist.

I replaced it with:

`<?= e(trans('backend::lang.form.create_title', ['name' => e(trans('rainlab.location::lang.state.label'))])) ?>`

This works!

**Comment by LukeTowers** [comment]
@LucasZdv are you still wanting to add the styling to the states lists widget?

**Comment by LucasZdv** [comment]
> @LucasZdv are you still wanting to add the styling to the states lists widget?

Yes, but it hasn't worked for me. Where should I do the extension? I added this to the controller, but not sure where it goes.

**Comment by LukeTowers** [comment]
In the Country controller. Do some debugging to see if the variables you're getting are what you expect.

**Comment by LucasZdv** [comment]
Works now!

```
    public function relationExtendViewWidget($widget)
    {
        $widget->bindEvent('list.injectRowClass', function ($record) {
            
            if(!$record->is_enabled) {
                return  'safe disabled';
            }
        });
    }
```
locations

File: controllers/Locations.php [hunk 14]
```diff
@@ -44,10 +44,21 @@ public function listInjectRowClass($record, $definition = null)
...
     }
 
+    public function relationExtendViewWidget($widget)
+    {
+        $widget->bindEvent('list.injectRowClass', function ($record) {
+            
+            if(!$record->is_enabled) {
+                return  'safe disabled';
+            }
+        });
```
Comment (LukeTowers) [inline controllers/Locations.php:54]:
```suggestion
        $widget->bindEvent('list.injectRowClass', function ($record) {
            if (!$record->is_enabled) {
                return  'safe disabled';
            }
        });
```

File: controllers/locations/_states_list_toolbar.htm [hunk 15]
```diff
@@ -0,0 +1,41 @@
+<div data-control="toolbar" data-disposable="">
+
```
Comment (LukeTowers) [inline controllers/locations/_states_list_toolbar.htm:2]:
```suggestion
```

**Comment by LucasZdv** [comment]
Hello, is it possible to go to master?

**Review by bennothommo — APPROVED** [review]

**Pull request merged**